src/Bicep.Tools.Benchmark/Compilation.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core;
+using Bicep.Core.Samples;
+using BenchmarkDotNet.Attributes;
+using Microsoft.Extensions.DependencyInjection;
+using System.Linq;
+using System.IO.Abstractions.TestingHelpers;
+using System.IO.Abstractions;
+using System.Collections.Immutable;
+using FluentAssertions;
+using System;
+
+namespace Bicep.Tools.Benchmark;
+
+public class Compilation
+{
+    private record BenchmarkData(
+        ImmutableArray<DataSet> DataSets,
+        IBicepService BicepService);
+
+    private static BenchmarkData CreateBenchmarkData()
+    {
+        var files = DataSet.ReadDataSetDictionary("Files");
+
+        var fileSystem = new MockFileSystem(files.ToDictionary(
+            x => x.Key,
+            x => new MockFileData(x.Value)));
+
+        var dataSets = DataSets.AllDataSets
+            .Where(x => !x.HasRegistryModules)
+            .ToImmutableArray();
+
+        var bicepService = new ServiceBuilder()
+            .WithRegistration(x => x.AddSingleton<IFileSystem>(fileSystem)).Build();
+
+        return new(dataSets, bicepService);
+    }
+
+    private BenchmarkData? benchmarkData;
+
+    [GlobalSetup]
+    public void GlobalSetup()
+    {
+        this.benchmarkData = CreateBenchmarkData();
+    }
+
+    [Benchmark(Description = "Get compilation diagnostics for each dataset")]
+    public void Get_compilation_diagnostics()
+    {
+        // Reuse a single IBicepService to amortize the cost of instantiating dependencies
+        var (dataSets, service) = benchmarkData!;
```

Comment:
>! [](http://example.com/codeflow?start=47&length=1)

Could probably get rid of the `!` by marking the member with the `[NotNull]` attribute like we do for TestContext in tests.